### PR TITLE
Add a more complete Closure to TypeScript type annotation converter.

### PR DESCRIPTION
**Pull request by aomarks**
Uses doctrine to get an AST from a type annotation. A more complete version of `getTsType` from `gen-ts.ts`. Still some TODOs, and not yet integrated into the generator.

### CODE REVIEW

File: src/closure-types.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * Convert from a type annotation in Closure syntax to TypeScript syntax (e.g
+ * `Array<string>` => `string[]|null`).
+ */
+export function closureTypeToTypeScript(closureType: string): string {
+  // Doctrine parses JSDoc annotations, and doesn't have an API purely for type
+  // expressions, so we'll wrap it in a JSDoc annotation and extract the part
+  // of the AST we care about.
+  const annotationAst = doctrine.parse('@return {' + closureType + '}');
```
Comment (justinfagnani) [inline src/closure-types.ts]: I think you can use this: https://github.com/eslint/doctrine/blob/master/lib/typed.js

File: src/closure-types.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    return '';
+  }
+
+  if (nullable === true || (nullable === null && nullableByDefault(node))) {
+    t = `${t}|null`;
+  }
+  return t;
+}
+
+function nullableByDefault(node: doctrine.Type): boolean {
```
Comment (justinfagnani) [inline src/closure-types.ts]: comment

File: src/closure-types.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+export function closureTypeToTypeScript(closureType: string): string {
+  // Doctrine parses JSDoc annotations, and doesn't have an API purely for type
+  // expressions, so we'll wrap it in a JSDoc annotation and extract the part
+  // of the AST we care about.
+  const annotationAst = doctrine.parse('@return {' + closureType + '}');
+  const typeAst = annotationAst.tags[0].type as doctrine.Type;
+  return serialize(typeAst);
+}
+
+function serialize(node: doctrine.Type): string {
```
Comment (justinfagnani) [inline src/closure-types.ts]: Add a comment and maybe add toTS to the name?

**Review by justinfagnani — APPROVED** [review]

File: src/closure-types.ts (hunk 1 above)
Comment (aomarks) [reply src/closure-types.ts]: Nice, done.

File: src/closure-types.ts (hunk 2 above)
Comment (aomarks) [reply src/closure-types.ts]: Done.

File: src/closure-types.ts (hunk 3 above)
Comment (aomarks) [reply src/closure-types.ts]: Done (comment).

**Pull request merged**